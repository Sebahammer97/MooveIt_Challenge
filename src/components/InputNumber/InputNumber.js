import React, { Component } from "react";
import cx from "classnames";
import PropTypes from "prop-types";
import NumberFormat from "react-number-format";

// Components
import Button from "../Button/Button";

class InputNumber extends Component {
  constructor(props) {
    super(props);
    this.state = { isOnLimit: false };
    this.handleInputChange.bind(this);
    this.numberLimits.bind(this);
  }

  // Definition of props types to this component
  static propTypes = {
    onChange: PropTypes.func,
    className: PropTypes.string,
    placeholder: PropTypes.string,
    value: PropTypes.string,
    variant: PropTypes.string,
    size: PropTypes.string,
    disabledClassName: PropTypes.string,
    disabled: PropTypes.bool,
    reset: PropTypes.bool,
    minValue: PropTypes.number,
    maxValue: PropTypes.number,
    format: PropTypes.string,
    mask: PropTypes.any,
  };

  // Assigning default values for the props
  static defaultProps = {
    className: "input",
    placeholder: "",
    value: "",
    variant: "default",
    disabled: false,
    disabledClassName: "",
    size: "medium",
    reset: false,
    mask: "",
  };

  handleInputChange = (formattedValue, value) => {
    const { onChange, disabled, minValue, maxValue } = this.props;

    if (disabled) {
      return;
    }

    if (onChange) {
      if (this.numberLimits(parseInt(value), minValue, maxValue)) {
        this.setState({ isOnLimit: true });
        onChange(formattedValue);
      } else {
        this.setState({ isOnLimit: false });
      }
    }
  };

  numberLimits = (number, minValue, maxValue) => {
    if (typeof minValue === "number") {
      if (minValue > number) {
        return false;
      }
    }
    if (typeof maxValue === "number") {
      if (maxValue < number) {
        return false;
      }
    }
    return true;
  };

  resetInputValue = () => {
    const { onChange } = this.props;
    if (onChange) {
      onChange("");
    }
  };

  render() {
    const {
      className,
      value,
      size,
      minValue,
      maxValue,
      variant,
      disabled,
      disabledClassName,
      reset,
      format,
      mask,
      placeholder,
      prefix,
      suffix,
    } = this.props;

    const styles = require("./style.module.css");

    const _className = cx(className, styles[size], styles[variant], {
      [styles.disabled]: disabled,
      [styles.error]:
        (minValue && parseInt(value) < parseInt(minValue)) ||
        (maxValue && parseInt(value) > parseInt(maxValue)),
      [disabledClassName]: disabled,
    });

    return (
      <div>
        <NumberFormat
          className={_className}
          value={value}
          allowLeadingZeros={false}
          thousandSeparator={true}
          isNumericString={false}
          onValueChange={({ formattedValue, value }) =>
            this.handleInputChange(formattedValue, value)
          }
          disabled={disabled}
          format={format}
          mask={mask}
          placeholder={placeholder}
          prefix={prefix}
          suffix={suffix}
          data-testid="input-number"
        />
        {reset && value.length > 0 ? (
          <Button label="X" variant="reset" onClick={this.resetInputValue} />
        ) : null}
      </div>
    );
  }
}

export default InputNumber;
