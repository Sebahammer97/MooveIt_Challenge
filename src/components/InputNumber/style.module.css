.input {
  position: relative;
  font-size: 14px;
  font-weight: 600;
  text-align: center;
  padding: 0.7em 1.2em;
  cursor: pointer;
  user-select: none;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  height: 32px;
  min-width: 96px;
  border-radius: 4px;
  background-color: #fff;
  border: 1px solid;
  color: #fff;
  margin-bottom: 0.5em;
  -webkit-transition: background 0.2s ease;
  -moz-transition: background 0.2s ease;
  -o-transition: background 0.2s ease;
  transition: background 0.2s ease;
}

.small {
  font-size: 8px;
}

.medium {
  font-size: 16px;
}

.large {
  font-size: 24px;
}

.default {
  background-color: #fff;
  border-color: #555;
  color: #000;
  border-radius: 10px;
  padding-left: 10px;
}

.default:focus {
  border-color: #61dafb;
}

.dark {
  background-color: #000;
  border-color: #555;
  color: #fff;
}

.disabled {
  background-color: #fff;
  border-color: #bababa;
  color: #bababa;
  box-shadow: none;
  opacity: 0.7;
  cursor: not-allowed;
  border: 1px solid;
}

.error {
  border: 1px solid #f00 !important;
}
