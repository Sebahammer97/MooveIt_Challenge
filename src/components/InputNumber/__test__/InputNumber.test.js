import React, { useState } from "react";
import ReactDOM from "react-dom";
import InputNumber from "../InputNumber";

import { render, cleanup, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";

afterEach(cleanup);

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<InputNumber />, div);
});

function InputNumberTest({
  minValue,
  maxValue,
  variant,
  disabled,
  format,
  mask,
  placeholder,
  prefix,
  suffix,
}) {
  const [input, setInput] = useState("");

  return (
    <InputNumber
      value={input}
      onChange={(value) => setInput(value)}
      variant={variant}
      disabled={disabled}
      minValue={minValue}
      maxValue={maxValue}
      format={format}
      mask={mask}
      placeholder={placeholder}
      prefix={prefix}
      suffix={suffix}
    />
  );
}

it("renders input number correctly", () => {
  const utils = render(<InputNumberTest />);
  const input = utils.getByTestId("input-number");
  fireEvent.change(input, { target: { value: "123" } });
  expect(input.value).toBe("123");
});

it("render input number format correctly", () => {
  const utils = render(
    <InputNumberTest
      format="####-####-####-####"
      mask="#"
      placeholder="####-####-####-####"
    />
  );
  const input = utils.getByTestId("input-number");
  fireEvent.change(input, { target: { value: "1234567890123456" } });
  expect(input.value).toBe("1234-5678-9012-3456");
});

it("render input number limit correctly", () => {
  const utils = render(<InputNumberTest minValue={0} maxValue={100} />);
  const input = utils.getByTestId("input-number");
  fireEvent.change(input, { target: { value: "55" } });
  expect(input.value).toBe("55");
  fireEvent.change(input, { target: { value: "999" } });
  expect(input.value).toBe("55");
});

it("render input number prefix correctly", () => {
  const utils = render(<InputNumberTest prefix="$" />);
  const input = utils.getByTestId("input-number");
  fireEvent.change(input, { target: { value: "100" } });
  expect(input.value).toBe("$100");
});

it("render input number suffix correctly", () => {
  const utils = render(<InputNumberTest suffix=" likes" />);
  const input = utils.getByTestId("input-number");
  fireEvent.change(input, { target: { value: "100" } });
  expect(input.value).toBe("100 likes");
});

it("matches snapshots input number", () => {
  const tree = renderer.create(<InputNumber />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("matches snapshots disabled", () => {
  const tree = renderer.create(<InputNumber disabled={true} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("matches snapshots with reset", () => {
  const tree = renderer.create(<InputNumber reset={true} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("matches snapshots size", () => {
  const tree = renderer
    .create(
      <div>
        <InputNumber size="small" /> <InputNumber size="medium" />{" "}
        <InputNumber size="large" />
      </div>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("matches snapshots variant styles", () => {
  const tree = renderer
    .create(
      <div>
        <InputNumber variant="dark" />
      </div>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("matches snapshots formats", () => {
  const tree = renderer
    .create(
      <div>
        <InputNumber
          format={"+54 (###) ####-####"}
          mask="_"
          placeholder="+54 (###) ####-####"
        />
        <InputNumber
          format={"####-####-####-####"}
          mask="#"
          placeholder="####-####-####-####"
        />
        <InputNumber
          format={"##/##/####"}
          mask={["D", "D", "M", "M", "Y", "Y", "Y", "Y"]}
          placeholder="DD/MM/YYYY"
        />
      </div>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("matches snapshots prefix", () => {
  const tree = renderer
    .create(
      <div>
        <InputNumber prefix="$" />
      </div>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("matches snapshots suffix", () => {
  const tree = renderer
    .create(
      <div>
        <InputNumber suffix=".-" />
      </div>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
