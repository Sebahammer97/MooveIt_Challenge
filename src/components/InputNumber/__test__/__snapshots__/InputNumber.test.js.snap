// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshots disabled 1`] = `
<div>
  <input
    className="input medium default disabled "
    data-testid="input-number"
    disabled={true}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseUp={[Function]}
    placeholder=""
    type="text"
    value=""
  />
</div>
`;

exports[`matches snapshots formats 1`] = `
<div>
  <div>
    <input
      className="input medium default"
      data-testid="input-number"
      disabled={false}
      inputMode="numeric"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseUp={[Function]}
      placeholder="+54 (###) ####-####"
      type="text"
      value=""
    />
  </div>
  <div>
    <input
      className="input medium default"
      data-testid="input-number"
      disabled={false}
      inputMode="numeric"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseUp={[Function]}
      placeholder="####-####-####-####"
      type="text"
      value=""
    />
  </div>
  <div>
    <input
      className="input medium default"
      data-testid="input-number"
      disabled={false}
      inputMode="numeric"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseUp={[Function]}
      placeholder="DD/MM/YYYY"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`matches snapshots input number 1`] = `
<div>
  <input
    className="input medium default"
    data-testid="input-number"
    disabled={false}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseUp={[Function]}
    placeholder=""
    type="text"
    value=""
  />
</div>
`;

exports[`matches snapshots prefix 1`] = `
<div>
  <div>
    <input
      className="input medium default"
      data-testid="input-number"
      disabled={false}
      inputMode="numeric"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseUp={[Function]}
      placeholder=""
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`matches snapshots size 1`] = `
<div>
  <div>
    <input
      className="input small default"
      data-testid="input-number"
      disabled={false}
      inputMode="numeric"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseUp={[Function]}
      placeholder=""
      type="text"
      value=""
    />
  </div>
   
  <div>
    <input
      className="input medium default"
      data-testid="input-number"
      disabled={false}
      inputMode="numeric"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseUp={[Function]}
      placeholder=""
      type="text"
      value=""
    />
  </div>
   
  <div>
    <input
      className="input large default"
      data-testid="input-number"
      disabled={false}
      inputMode="numeric"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseUp={[Function]}
      placeholder=""
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`matches snapshots suffix 1`] = `
<div>
  <div>
    <input
      className="input medium default"
      data-testid="input-number"
      disabled={false}
      inputMode="numeric"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseUp={[Function]}
      placeholder=""
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`matches snapshots variant styles 1`] = `
<div>
  <div>
    <input
      className="input medium dark"
      data-testid="input-number"
      disabled={false}
      inputMode="numeric"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseUp={[Function]}
      placeholder=""
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`matches snapshots with reset 1`] = `
<div>
  <input
    className="input medium default"
    data-testid="input-number"
    disabled={false}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseUp={[Function]}
    placeholder=""
    type="text"
    value=""
  />
</div>
`;
